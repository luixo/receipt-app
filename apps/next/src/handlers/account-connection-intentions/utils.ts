import * as trpc from "@trpc/server";

import { Database } from "next-app/db";
import { ACCOUNT_CONNECTIONS_INTENTIONS } from "next-app/db/consts";
import { AccountsId, UsersId } from "next-app/db/models";
import { getUserById } from "next-app/handlers/users/utils";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: string,
	asUserId: UsersId
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select("id")
		.where("email", "=", toEmail)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Account with email ${toEmail} does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId)
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new trpc.TRPCError({
			code: "CONFLICT",
			message: `Account with email ${toEmail} is already connected to user ${connectedUser.name}.`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where("accountId", "=", targetAccount.id)
		.where("targetAccountId", "=", fromAccountId)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where("ownerAccountId", "=", targetAccount.id)
				.where("id", "=", viceVersaIntention.userId)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where("ownerAccountId", "=", fromAccountId)
				.where("id", "=", asUserId)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where("accountId", "=", targetAccount.id)
				.where("targetAccountId", "=", fromAccountId)
				.executeTakeFirst();
		});
		return {
			id: targetAccount.id,
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				created: new Date(),
			})
			.executeTakeFirst();
		return {
			id: targetAccount.id,
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where("accountId", "=", fromAccountId)
				.where("targetAccountId", "=", targetAccount.id)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await getUserById(
				database,
				existingIntention.userId,
				["name"]
			);
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to ${toEmail} as user ${
					existingUser!.name
				}.`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user ${user.name}.`,
			});
		}
		throw e;
	}
};

export const removeIntention = async (
	database: Database,
	intention:
		| {
				targetAccountId: AccountsId;
				accountId: AccountsId;
		  }
		| undefined,
	intentionType: string
) => {
	if (!intention) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Intention ${intentionType} does not exist.`,
		});
	}
	await database
		.deleteFrom("accountConnectionsIntentions")
		.where("accountId", "=", intention.accountId)
		.where("targetAccountId", "=", intention.targetAccountId)
		.execute();
};
