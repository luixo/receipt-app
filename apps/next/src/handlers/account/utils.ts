import { Selection } from "kysely";
import { Database, ReceiptsSelectExpression, ReceiptsDatabase } from "../../db";
import { AccountsId } from "../../db/models";

export const getAccountById = <SE extends ReceiptsSelectExpression<"accounts">>(
	database: Database,
	id: AccountsId,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("id", "=", id)
		.executeTakeFirst();
};

export const getAccountByEmail = <
	SE extends ReceiptsSelectExpression<"accounts">
>(
	database: Database,
	email: string,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("email", "=", email)
		.executeTakeFirst();
};
