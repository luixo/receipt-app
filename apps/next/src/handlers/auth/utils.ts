import * as trpc from "@trpc/server";

import { DAY } from "app/utils/time";
import type { Database } from "next-app/db";
import type { AccountsId, SessionsSessionId } from "next-app/db/models";
import { generateConfirmEmailEmail } from "next-app/email/utils";
import { getUuid } from "next-app/utils/crypto";
import { getEmailClient, isEmailServiceActive } from "next-app/utils/email";

// How long a session should last
const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long before the session expires should we update the expiration date
const SESSION_SHOULD_UPDATE_DURATION = 28 * DAY;

const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);
export const shouldUpdateExpirationDate = (expirationTimestamp: Date) =>
	expirationTimestamp.valueOf() - Date.now() < SESSION_SHOULD_UPDATE_DURATION;

export const createAuthorizationSession = async (
	database: Database,
	accountId: AccountsId,
) => {
	const uuid = getUuid();
	const expirationDate = getExpirationDate();
	await database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const updateAuthorizationSession = async (
	database: Database,
	authToken: SessionsSessionId,
) => {
	database
		.updateTable("sessions")
		.set({ expirationTimestamp: getExpirationDate() })
		.where("sessionId", "=", authToken)
		.executeTakeFirst();
};

export const removeAuthorizationSession = async (
	database: Database,
	sessionId: SessionsSessionId,
) => {
	await database
		.deleteFrom("sessions")
		.where("sessionId", "=", sessionId)
		.executeTakeFirst();
};

export const sendVerificationEmail = async (email: string, token: string) => {
	if (!isEmailServiceActive()) {
		throw new trpc.TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient().send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(token),
		});
	} catch (e) {
		throw new trpc.TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
