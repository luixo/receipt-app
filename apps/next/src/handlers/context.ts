import * as trpcNext from "@trpc/server/adapters/next";
import { NextApiRequest, NextApiResponse } from "next";
import { Logger, logger } from "../utils/logger";

export type Context = {
	req: NextApiRequest;
	res: NextApiResponse;
	logger: Logger;
	debug: boolean;
};

export const createContext = (
	opts: trpcNext.CreateNextContextOptions
): Context => ({
	req: opts.req,
	res: opts.res,
	logger,
	debug: Boolean(opts.req.headers.debug),
});
