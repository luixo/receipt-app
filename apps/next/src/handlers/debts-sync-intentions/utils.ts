import * as trpc from "@trpc/server";
import { z } from "zod";

import { Database } from "next-app/db";
import { AccountsId, DebtsId } from "next-app/db/models";
import { getValidParticipants } from "next-app/handlers/receipt-items/utils";

export const getDebtIntention = async (
	database: Database,
	debtId: DebtsId,
	accountId: AccountsId,
) =>
	database
		.selectFrom("debts")
		.where("debts.id", "=", debtId)
		.leftJoin("debts as debtsMine", (qb) =>
			qb
				.onRef("debts.id", "=", "debtsMine.id")
				.on("debtsMine.ownerAccountId", "=", accountId),
		)
		.leftJoin("debtsSyncIntentions", (qb) =>
			qb.onRef("debts.id", "=", "debtsSyncIntentions.debtId"),
		)
		.leftJoin("debts as debtsTheir", (qb) =>
			qb
				.onRef("debts.id", "=", "debtsTheir.id")
				.on("debtsTheir.ownerAccountId", "<>", accountId),
		)
		.select([
			"debtsSyncIntentions.ownerAccountId as intentionAccountId",
			"debtsMine.lockedTimestamp as mineLockedTimestamp",
			"debtsMine.ownerAccountId as mineOwnerAccountId",
			"debtsTheir.lockedTimestamp as theirLockedTimestamp",
			"debtsTheir.ownerAccountId as theirOwnerAccountId",
		])
		.executeTakeFirst();

const syncStatusSchema = z.discriminatedUnion("type", [
	// The foreign user doesn't have a debt
	z.strictObject({ type: z.literal("nosync") }),
	// The foreign user debt's data is different
	z.strictObject({
		type: z.literal("unsync"),
		intention: z
			.strictObject({
				direction: z.union([z.literal("self"), z.literal("remote")]),
			})
			.optional(),
	}),
	// The foreign user debt's data is in sync
	z.strictObject({ type: z.literal("sync") }),
]);

export type SyncStatus = z.infer<typeof syncStatusSchema>;

export const getSyncStatus = async (
	database: Database,
	debtId: DebtsId,
	accountId: AccountsId,
): Promise<SyncStatus> => {
	const debtIntention = await getDebtIntention(database, debtId, accountId);
	if (!debtIntention) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Debt ${debtId} does not exist.`,
		});
	}
	const {
		mineOwnerAccountId,
		mineLockedTimestamp,
		theirOwnerAccountId,
		theirLockedTimestamp,
		intentionAccountId,
	} = debtIntention;
	if (!mineOwnerAccountId) {
		throw new trpc.TRPCError({
			code: "FORBIDDEN",
			message: `Debt ${debtId} does not exist for user ${accountId}.`,
		});
	}
	if (
		theirLockedTimestamp &&
		mineLockedTimestamp &&
		theirLockedTimestamp.valueOf() === mineLockedTimestamp.valueOf()
	) {
		return { type: "sync" };
	}
	if (!intentionAccountId) {
		if (theirOwnerAccountId) {
			return { type: "unsync" };
		}
		return { type: "nosync" };
	}
	if (intentionAccountId === accountId) {
		return {
			type: "unsync",
			intention: { direction: "self" },
		};
	}
	return {
		type: "unsync",
		intention: { direction: "remote" },
	};
};

type Participant = Awaited<ReturnType<typeof getValidParticipants>>[number];

export const upsertDebtIntentionFromReceipt = async (
	database: Database,
	participants: Participant[],
	ownerAccountId: AccountsId,
	created: Date,
) =>
	database
		.insertInto("debtsSyncIntentions")
		.values(
			participants.map((participant) => ({
				debtId: participant.debtId,
				ownerAccountId,
				lockedTimestamp: created,
			})),
		)
		.onConflict((oc) =>
			oc.column("debtId").doUpdateSet({
				ownerAccountId: (eb) => eb.ref("excluded.ownerAccountId"),
				lockedTimestamp: (eb) => eb.ref("excluded.lockedTimestamp"),
			}),
		)
		.execute();
