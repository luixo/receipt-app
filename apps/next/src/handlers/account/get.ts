import * as trpc from "@trpc/server";
import { z } from "zod";

import { AuthorizedContext } from "../context";
import { getDatabase } from "../../db";
import { TRPCError } from "@trpc/server";

export const router = trpc.router<AuthorizedContext>().query("get", {
	input: z.undefined(),
	resolve: async ({ ctx }) => {
		const database = getDatabase(ctx);
		const account = await database
			.selectFrom("accounts")
			.innerJoin("users", (jb) =>
				jb.onRef("users.ownerAccountId", "=", "accounts.id")
			)
			.select(["accounts.id", "users.name"])
			.where("accounts.id", "=", ctx.auth.accountId)
			.executeTakeFirst();
		if (!account) {
			throw new TRPCError({
				code: "NOT_FOUND",
				message: `Account ${ctx.auth.accountId} is not found`,
			});
		}
		return account;
	},
});
