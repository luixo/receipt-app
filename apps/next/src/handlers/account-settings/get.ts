import type { ReceiptsDatabase } from "next-app/db/types";
import { authProcedure } from "next-app/handlers/trpc";

type SettingsKey = Omit<ReceiptsDatabase["accountSettings"], "accountId">;
type Settings = {
	[K in keyof SettingsKey]: SettingsKey[K]["__select__"];
};

export const DEFAULT_ACCOUNT_SETTINGS: Settings = {
	autoAcceptDebts: false,
};

export const procedure = authProcedure.query(async ({ ctx }) => {
	const { database } = ctx;
	const account = await database
		.selectFrom("accountSettings")
		.select(["accountSettings.autoAcceptDebts"])
		.where("accountSettings.accountId", "=", ctx.auth.accountId)
		.executeTakeFirst();
	return account || DEFAULT_ACCOUNT_SETTINGS;
});
