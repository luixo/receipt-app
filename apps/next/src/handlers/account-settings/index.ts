import { t } from "next-app/handlers/trpc";

import { procedure as get } from "./get";
import { procedure as update } from "./update";

export const router = t.router({
	get,
	update,
});
