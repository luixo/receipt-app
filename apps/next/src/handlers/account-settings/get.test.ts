import { describe, expect } from "vitest";

import { t } from "next-app/handlers/trpc";
import { createAuthContext } from "next-tests/utils/context";
import {
	insertAccountSettings,
	insertAccountWithSession,
} from "next-tests/utils/data";
import { expectUnauthorizedError } from "next-tests/utils/expect";
import { test } from "next-tests/utils/test";

import { procedure } from "./get";

const router = t.router({ procedure });

describe("accountSettings.get", () => {
	describe("input verification", () => {
		expectUnauthorizedError((context) =>
			router.createCaller(context).procedure(),
		);
	});

	describe("functionality", () => {
		test("settings not found - default returned", async ({ ctx }) => {
			const { sessionId } = await insertAccountWithSession(ctx);
			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			await expect(await caller.procedure()).toMatchSnapshot();
		});

		test("settings found - default", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			await insertAccountSettings(ctx, accountId, { autoAcceptDebts: false });
			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			await expect(await caller.procedure()).toMatchSnapshot();
		});

		test("settings found - changed", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			await insertAccountSettings(ctx, accountId, { autoAcceptDebts: true });
			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			await expect(await caller.procedure()).toMatchSnapshot();
		});
	});
});
