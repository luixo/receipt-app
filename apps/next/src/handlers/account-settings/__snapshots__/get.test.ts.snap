// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`accountSettings.get > functionality > settings found - changed 1`] = `
{
  "autoAcceptDebts": true,
}
`;

exports[`accountSettings.get > functionality > settings found - default 1`] = `
{
  "autoAcceptDebts": false,
}
`;

exports[`accountSettings.get > functionality > settings not found - default returned 1`] = `
{
  "autoAcceptDebts": false,
}
`;
